src/components/users/controllers.tsx

```diff
@@ -95,3 +106,208 @@ export async function getUser(
+      body: template.render(<PasswordResetRequest csrf={ctx.viewContext.csrf} invalidEmail={true} />),
+      status: 400,
+    };
+  }
+
+  const uaa = new UAAClient({
+    apiEndpoint: ctx.app.uaaAPI,
+    clientCredentials: {
+      clientID: ctx.app.oauthClientID,
+      clientSecret: ctx.app.oauthClientSecret,
+    },
+  });
+
+  const accounts = new AccountsClient({
+    apiEndpoint: ctx.app.accountsAPI,
+    logger: ctx.app.logger,
+    secret: ctx.app.accountsSecret,
+  });
+
+  const accountsUser = await accounts.getUserByEmail(email);
+  if (!accountsUser) {
+    // 100ms delay to stop people iterating through users
+    await new Promise(finish => setTimeout(finish, 100));
+
+    template.title = 'Error: User not found';
+
+    return {
+      body: template.render(<PasswordResetRequest csrf={ctx.viewContext.csrf} userNotFound={true} />),
+      status: 404,
+    };
+  }
+
+  const uaaUser = await uaa.getUser(accountsUser.uuid)
+  if (!uaaUser) {
+    // 100ms delay to stop people iterating through users
+    await new Promise(finish => setTimeout(finish, 100));
+
+    template.title = 'Error: User not found';
+
+    return {
+      body: template.render(<PasswordResetRequest csrf={ctx.viewContext.csrf} userNotFound={true} />),
+      status: 404,
+    };
+  }
+
+  if (uaaUser.origin != 'uaa') {
+    // 100ms delay to stop people iterating through users
+    await new Promise(finish => setTimeout(finish, 100));
+
+    const idpNice = capitalize(uaaUser.origin);
+    template.title = `Error: You have enabled ${idpNice} single sign-on`;
+
+    return {
+      body: template.render(<PasswordResetRequest csrf={ctx.viewContext.csrf} userEnabledSSO={true} idpNice={idpNice}/>),
+    };
+  }
+
+  const notify = new NotificationClient({
+    apiKey: ctx.app.notifyAPIKey,
+    templates: {
+      passwordReset: ctx.app.notifyPasswordResetTemplateID,
+    },
+  });
+  const resetCode = await uaa.obtainPasswordResetCode(uaaUser.userName);
+
+  const url = new URL(ctx.app.domainName);
+  url.pathname = '/password/confirm-reset';
+  url.searchParams.set('code', resetCode);
+
+  await notify.sendPasswordReminder(email, url.toString());
+
+  template.title = 'Password reset successfully requested';
+  const successPage = <PasswordResetSuccess
+    title="Password reset successfully requested"
+    message="You should receive an email shortly with further instructions."
+  />;
+
+  return {
+    body: template.render(successPage),
+  };
+}
+
+export async function resetPasswordProvideNew(ctx: IContext, params: IParameters): Promise<IResponse> {
+  if (!params.code) {
+    throw new NotFoundError('Invalid password reset token.');
+  }
+
+  const template = new Template(ctx.viewContext, 'Password reset');
+
+  return await Promise.resolve({
+    body: template.render(<PasswordResetSetPasswordForm csrf={ctx.viewContext.csrf} code={params.code} />),
+  });
+}
+
+export function checkPasswordAgainstPolicy(pw: string): { valid: boolean; message?: string } {
```

Comment:
```suggestion
export function checkPasswordAgainstPolicy(pw: string): { valid: boolean; messages?: ReadonlyArray<string> } {
```

That way, user won't have to re-submit the form multiple times before they get their password right.

Reply:
@kr8n3r is it cool to display all the messages at the same time? if so I'll introduce this change